### PR TITLE
Make a non-finite training step visible in log_history

### LINKED ISSUE #6702: GRPO: a NaN training step is invisible in log_history, so a corrupted run reports a clean loss curve
### Summary

A training step whose loss is NaN is not visible in `log_history`. The reported loss curve stays finite and plausible while the step that produced it was NaN, so a run that is silently corrupting its weights looks healthy in the logs and in any tracker fed from them.

This came out of the regression test for #6166 (PR #6693). @albertvillanova asked for it to be tracked separately, since it is orthogonal to that fix.

### Mechanism

`transformers.Trainer` filters NaN and Inf out of the logged running loss by design:

```python
# transformers/trainer.py:1746
if (
    self.args.logging_nan_inf_filter
    and not is_torch_xla_available()
    and (torch.isnan(tr_loss_step) or torch.isinf(tr_loss_step))
):
    # if loss is nan or inf simply add the average of previous logged losses
    self._tr_loss += self._tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)
else:
    ...
```

`logging_nan_inf_filter` defaults to `True` (`transformers/training_args.py:977`), and its own help text says it "Does not affect gradient computation, only logging". So the NaN step still runs `backward()` and still takes an optimizer step. Only the *report* is repaired, by substituting the mean of the previously logged losses.

### Why this matters more for GRPO than for plain SFT

The GRPO loss has several paths that can produce a NaN from data rather than from divergence: an unscorable token logprob coming back from vLLM (#6166), a degenerate reward group, an overflowing `exp` of a KL log-ratio (#3015). In those cases the NaN is not a symptom of a bad learning rate that a user would go looking for. It is a single poisoned token or group, and the resulting NaN gradients propagate into the weights while the loss curve shows nothing at all.

### Reproduction

Observed while building the #6166 test. With the fix removed, an individual step's loss is NaN at the point `_compute_loss` returns it, and `state.log_history["train_loss"]` is still finite for that same run. That is why the test in PR #6693 asserts on the losses returned by `_compute_loss` rather than on `log_history`: asserting on `log_history` passes even with the fix removed, which makes it a vacuous test.

### Possible directions

Listing these as options rather than a proposal, since the right layer is a maintainer call:

1. Count NaN and Inf steps and surface the count as a metric, so the filtering stays but stops being silent.
2. Emit a warning the first time a step is filtered, naming the step index.
3. Leave `transformers` alone and check for a non-finite loss inside the GRPO loss path, where the cause is known and can be named.

Option 3 keeps the change inside TRL and can point at the actual source, which the generic filter cannot do.

I am happy to implement whichever direction you prefer.

cc @qgallouedec @kashif

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
The training logger could hide a non-finite loss while backward still used it. I added a gathered `frac_nonfinite_loss` metric, warnings, and agreement across ranks for selected failure paths to prevent peers from hanging. Regression tests expose the missing metric on main, and distributed GPU checks pass for the forward-failure agreement, while local async tests remain blocked by an unavailable kernel build.
Resolves #6702.

---

> [!NOTE]
> **Medium Risk**
> Touches every trainer’s hot loss path with new collectives and cross-rank failure agreement; incorrect placement could cause deadlocks or change logging semantics, though behavior is heavily tested.
> 
> **Overview**
> Adds **`frac_nonfinite_loss`** across TRL trainers so NaN/Inf steps are visible in logs even when Transformers’ default **`logging_nan_inf_filter`** keeps the reported **`loss`** finite. Each loss path records a rank-gathered fraction of non-finite micro-batch losses, emits **train/eval** **`warning_once`** messages, and documents the metric on trainer pages.
> 
> Also fixes eval logging: metrics use the caller’s **`metric_key_prefix`** (not a hardcoded **`eval_`**) and **`evaluate`/`predict`** open a clean eval metrics window so **`predict()`** batches do not leak into the next eval average. **SFT** **`evaluate`** on a dict of splits prepares each split once instead of double-preparing via nested **`Trainer.evaluate`**.
> 
> For **DPO**, **KTO**, and **SFT**, forward failures are agreed across ranks **before** metric gathers so one rank’s exception does not hang peers in collectives; distributed child tests cover gathered **`frac_nonfinite_loss`** and peer-failure raises. Broad unit tests poison losses (NaN/Inf where applicable) and assert **`frac_nonfinite_loss`** behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f0961196153df7493d1da6767c304a26d6f49f15. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/async_grpo/async_grpo_trainer.py [hunk 1]
```diff
@@ -1234,6 +1237,25 @@ def seg_sum(vals):  # per-completion segment sum over the packed row
...
+            logger.warning_once(
+                "The training loss is not finite (NaN or Inf) for at least one step. The logged loss may not show it: "
+                "when `logging_nan_inf_filter` is enabled, which is the default, and `is_torch_xla_available()` is "
+                "false, `transformers` discards the step's own loss and logs a value derived from the losses "
+                "accumulated since the last training log. The backward pass still runs on the non-finite value; "
+                "whether that ultimately produces a parameter update depends on the resulting gradients and the "
+                "configured scaler, optimizer and backend. `frac_nonfinite_loss` reports the fraction of ranks whose "
+                "loss was non-finite, averaged over every loss computation since the last log, so it is a rate rather "
+                "than a count of affected optimizer steps."
+            )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/async_grpo/async_grpo_trainer.py]: Correct for the commit this review ran on, and already fixed one commit later.

Bugbot reviewed `2d67e00c`. At that commit `async_grpo_trainer.py:1247` really did read
`self._metrics["train"]["frac_nonfinite_loss"].append(...)` with the sink hardcoded, and it emitted only
the training message, so a non-finite evaluation batch was filed as a training rate and the evaluation
warning never fired.

`6eb4686a` replaced that with the mode branch:

```python
mode = "train" if self.model.training else "eval"
nonfinite = self.accelerator.gather((~torch.isfinite(loss.detach().mean())).float())
self._metrics[mode]["frac_nonfinite_loss"].append(nonfinite.mean().item())
```

followed by a `train`/`eval` split on the warning text. `git log -S` puts that change in `6eb4686a`. The
guard is now byte-identical to `AsyncDistillationTrainer`, the sibling this review cited as the
reference: the drift checker compares all 30 copies of the block across the repository and reports zero
divergence.

No code change needed on the current head.

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -1753,32 +1771,103 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+            else:
+                forward_error = e
+        failed_anywhere = torch.tensor(float(forward_error is not None), device=self.accelerator.device)
+        if self.accelerator.gather(failed_anywhere).any():
+            if forward_error is not None:
+                raise forward_error
+            raise RuntimeError(
+                "The forward pass failed on another rank. This rank raises too so the run fails together instead of "
+                "blocking in the next collective; the cause is in the failing rank's traceback."
+            )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/dpo_trainer.py]: Confirmed, and it was worse than the placement of one gather. `_compute_loss` runs sixteen metric gathers between its forward and its return, so an agreement in `compute_loss` sat below all of them: the survivor's first metric gather paired with the failing rank's agreement flag and the run blocked in the next collective.

Fixed in 9ac794a9. The agreement now sits inside `_compute_loss` and `_compute_loss_liger`, directly after the forward and before any collective, in both DPO and KTO. KTO's reference forward runs below the batch-size check's gather, so it carries an agreement of its own. `compute_loss` is plain delegation again and the non-finite guard moved into the callees, so it stays the last collective on every path.

822f1c4c widens the catch from `ValueError` to `Exception`, so a `RuntimeError` or an out-of-memory error on one rank reaches the same agreement instead of leaving the others waiting, and gives KTO's fused loss an agreement of its own, since it sits downstream of the KL gather.

Measured with `tests/distributed/rank_local_forward_failure.py` on two ranks: the shipped code makes every rank raise; with both agreements removed the survivor stalls until the watchdog ends it. The child now fails the forward from a pre-hook rather than raising in an override, since an override that raised before delegating would skip the agreement it exists to test, and it refuses to run on one rank, where every assertion passes either way.

File: tests/test_dpo_trainer.py [hunk 3]
```diff
@@ -378,7 +379,105 @@ def test_train_dataset_format(self, config_name):
...
+
+                    def poison_logits(module, args, output):
+                        output.logits = output.logits * float("nan")
+                        return output
+
+                    handle = model.register_forward_hook(poison_logits)
+                    try:
+                        return super()._compute_loss_liger(model, inputs, return_outputs)
+                    finally:
+                        handle.remove()
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply tests/test_dpo_trainer.py]: Confirmed. `_compute_loss_liger` calls the backbone and reads `last_hidden_state`; it never calls the wrapper the hook was registered on and never reads `logits`, so the hook did not fire. CI agreed: the test failed with `assert 0.0 == 1.0` in three lanes on 0cab4014.

Fixed in 9ac794a9. The test wraps `self.liger_loss` and adds the poison to the loss it returns, which is the value the Liger guard reads. Because the poison lands on the loss itself, an injected `inf` survives as `inf` rather than collapsing to NaN, so the test is now parametrized over both.

File: trl/trainer/sft_trainer.py [hunk 4]
```diff
@@ -1745,6 +1752,21 @@ def evaluate(
...
+            for name, dataset in eval_dataset.items():
+                # Each split is its own evaluation window, and `super().evaluate` does not re-enter this override, so
+                # open the window per split with that split's prefix.
+                self._open_eval_window(f"{metric_key_prefix}_{name}")
+                metrics.update(
+                    super().evaluate(
+                        eval_dataset=dataset, ignore_keys=ignore_keys, metric_key_prefix=f"{metric_key_prefix}_{name}"
+                    )
+                )
+            return metrics
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/sft_trainer.py:1773]: Measured on this head (4db11dba, transformers 5.11.0) with `_prepare_dataset` wrapped in a call counter, `eval_dataset={"a": zen_test, "b": zen_test}` passed at init, then `trainer.evaluate()` with no argument:

```
packing=False  prepare_calls: init=3  evaluate()=0  evaluate(dict)=2  rows before/after evaluate(): {'a': 2, 'b': 2} -> {'a': 2, 'b': 2}
packing=True   prepare_calls: init=3  evaluate()=0  evaluate(dict)=2  rows before/after evaluate(): {'a': 1, 'b': 1} -> {'a': 1, 'b': 1}
```

The no-argument call prepares nothing and re-packs nothing. The reason is the `override` flag in `Trainer.evaluate` (transformers/trainer.py:2581-2587, same shape at 4.56.2): when `eval_dataset` is `None` it recurses with `eval_dataset=<split name>`, a `str`, and this override leaves `str` inputs untouched (the `not isinstance(eval_dataset, str)` guard at the top of the preparation block). Only the explicit-dict argument path prepares, once per split, which is the case this PR's loop handles.

**Comment by qgallouedec** [comment]
Thanks @behroozazarkhalili but I don't think I get the purpose here: we're adding a metric that only lights up once the run is already dead? In what real training did you end up finding this metric really useful?
Also the description is way too long and reads as AI generated, It would benefit from being cut it to a few sentences

**Comment by behroozazarkhalili** [comment]
You are right on both points.

There was no real run. This came from the regression test for #6166. An assertion on `log_history` passed with the fix removed, because the NaN step had been replaced by the running mean. I have not seen it in an actual training job, and every NaN in this PR's tests is injected.

I also agree on the timing. Without a grad scaler, a NaN loss still takes an optimizer step, so the run is already dead by the time anything is logged. `grad_norm` is logged unfiltered (`trainer.py:2081`), so the next log line shows `nan` there anyway. The metric adds little on top of that. It only tells you that the finite loss column is no longer real. The one case where it could tell you something `grad_norm` does not is fp16 with a grad scaler, where the step is skipped and training continues. I have not measured that case, so I will not rely on it.

Given that, 70 files is far more than this is worth, and I am fine closing it. If you see value in a small version, it would be a warning on the first filtered step in GRPO and RLOO only. It would have to be a new PR from my fork, because this branch now conflicts with #7136 and I can no longer push to it. It is your decision.

Noted on the description. I will keep descriptions to a few sentences from now on.

**Comment by qgallouedec** [comment]
Closing, I think it's important to work from real training. And fix real issue. The amount of edge / unreachable cases that we can find from just reading the codebase is infinite.

**Comment by behroozazarkhalili** [comment]
That's a fair bar, and I agree with it.

Six of my other open PRs had the same origin. I found them by reading the code, not because a run broke or a user reported a problem. I've closed all six (#6977, #7033, #7042, #7044, #7045, #7053) so they're out of your queue. I also closed the issues I had filed for them.

From here on, I'll only open a PR when there's a failing run or a user report behind it, and the description will start with that evidence. I appreciate the direct feedback.
